File: qc-storage/src/main/java/com/eager/questioncloud/storage/subscribe/SubscribeRepositoryImpl.java
package com.eager.questioncloud.storage.subscribe;

import static com.eager.questioncloud.storage.creator.QCreatorEntity.creatorEntity;
import static com.eager.questioncloud.storage.creator.QCreatorStatisticsEntity.creatorStatisticsEntity;
import static com.eager.questioncloud.storage.subscribe.QSubscribeEntity.subscribeEntity;
import static com.eager.questioncloud.storage.user.QUserEntity.userEntity;

import com.eager.questioncloud.core.common.PagingInformation;
import com.eager.questioncloud.core.domain.creator.dto.CreatorDto.CreatorInformation;
import com.eager.questioncloud.core.domain.subscribe.model.Subscribe;
import com.eager.questioncloud.core.domain.subscribe.repository.SubscribeRepository;
import com.querydsl.core.types.Projections;
import com.querydsl.jpa.impl.JPAQueryFactory;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class SubscribeRepositoryImpl implements SubscribeRepository {
    private final SubscribeJpaRepository subscribeJpaRepository;
    private final JPAQueryFactory jpaQueryFactory;

    @Override
    public Subscribe save(Subscribe subscribe) {
        return subscribeJpaRepository.save(SubscribeEntity.from(subscribe)).toModel();
    }

    @Override
    public Boolean isSubscribed(Long subscriberId, Long creatorId) {
        return subscribeJpaRepository.existsBySubscriberIdAndCreatorId(subscriberId, creatorId);
    }

    @Override
    public void unSubscribe(Long subscriberId, Long creatorId) {
        subscribeJpaRepository.deleteBySubscriberIdAndCreatorId(subscriberId, creatorId);
    }

    @Override
    public int countSubscriber(Long creatorId) {
        return subscribeJpaRepository.countByCreatorId(creatorId);
    }

    @Override
    public List<CreatorInformation> getSubscribeCreators(Long userId, PagingInformation pagingInformation) {
        return jpaQueryFactory.select(
                Projections.constructor(
                    CreatorInformation.class,
                    creatorEntity.id,
                    userEntity.userInformationEntity.name,
                    userEntity.userInformationEntity.profileImage,
                    creatorEntity.creatorProfileEntity.mainSubject,
                    userEntity.userInformationEntity.email,
                    creatorStatisticsEntity.salesCount,
                    creatorStatisticsEntity.averageRateOfReview,
                    creatorEntity.creatorProfileEntity.introduction))
            .from(subscribeEntity)
            .where(subscribeEntity.subscriberId.eq(userId))
            .leftJoin(creatorEntity).on(creatorEntity.id.eq(subscribeEntity.creatorId))
            .leftJoin(userEntity).on(userEntity.uid.eq(creatorEntity.userId))
            .leftJoin(creatorStatisticsEntity).on(creatorStatisticsEntity.creatorId.eq(creatorEntity.id))
            .offset(pagingInformation.getOffset())
            .limit(pagingInformation.getSize())
            .fetch();
    }

    @Override
    public int countMySubscribe(Long userId) {
        Integer count = jpaQueryFactory.select(subscribeEntity.id.count().intValue())
            .from(subscribeEntity)
            .where(subscribeEntity.subscriberId.eq(userId))
            .fetchFirst();

        if (count == null) {
            return 0;
        }

        return count;
    }
}

File: qc-community/src/main/kotlin/com/eager/questioncloud/subscribe/infrastructure/repository/SubscribeRepositoryImpl.kt
package com.eager.questioncloud.subscribe.infrastructure.repository

import com.eager.questioncloud.common.pagination.PagingInformation
import com.eager.questioncloud.subscribe.domain.Subscribe
import com.eager.questioncloud.subscribe.infrastructure.entity.QSubscribeEntity.subscribeEntity
import com.eager.questioncloud.subscribe.infrastructure.entity.SubscribeEntity
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
class SubscribeRepositoryImpl(
    private val subscribeJpaRepository: SubscribeJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory,
) : SubscribeRepository {
    override fun save(subscribe: Subscribe): Subscribe {
        return subscribeJpaRepository.save(SubscribeEntity.from(subscribe)).toModel()
    }

    override fun isSubscribed(subscriberId: Long, creatorId: Long): Boolean {
        return subscribeJpaRepository.existsBySubscriberIdAndCreatorId(subscriberId, creatorId)
    }

    @Transactional
    override fun unSubscribe(subscriberId: Long, creatorId: Long) {
        subscribeJpaRepository.deleteBySubscriberIdAndCreatorId(subscriberId, creatorId)
    }

    override fun countSubscriber(creatorId: Long): Int {
        return subscribeJpaRepository.countByCreatorId(creatorId)
    }

    override fun countSubscriber(creatorIds: List<Long>): Map<Long, Int> {
        return jpaQueryFactory.select(subscribeEntity.creatorId, subscribeEntity.id.count().intValue())
            .from(subscribeEntity)
            .where(subscribeEntity.creatorId.`in`(creatorIds))
            .groupBy(subscribeEntity.creatorId)
            .fetch()
            .associate { it.get(subscribeEntity.creatorId)!! to it.get(subscribeEntity.id.count().intValue())!! }
    }

    override fun getMySubscribedCreators(userId: Long, pagingInformation: PagingInformation): List<Long> {
        return jpaQueryFactory.select(subscribeEntity.creatorId)
            .from(subscribeEntity)
            .where(subscribeEntity.subscriberId.eq(userId))
            .offset(pagingInformation.offset.toLong())
            .limit(pagingInformation.size.toLong())
            .fetch()
    }

    override fun countMySubscribe(userId: Long): Int {
        return jpaQueryFactory.select(subscribeEntity.id.count().intValue())
            .from(subscribeEntity)
            .where(subscribeEntity.subscriberId.eq(userId))
            .fetchFirst() ?: 0
    }

    override fun deleteAllInBatch() {
        subscribeJpaRepository.deleteAllInBatch()
    }
}
